Solutions/Marain.ContentManagement.Api.Services.AspNetCore/Microsoft/Extensions/DependencyInjection/ContentManagementServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,102 @@
+
+    /// <summary>
+    /// Extension methods for configuring DI for the Content Management Open API services.
+    /// </summary>
+    public static class ContentManagementServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Add services required by the content management API.
+        /// </summary>
+        /// <param name="services">The service collection.</param>
+        /// <param name="rootTenantDefaultConfiguration">
+        /// Configuration section to read root tenant default repository settings from.
+        /// </param>
+        /// <param name="configureHost">Optional callback for additional host configuration.</param>
+        /// <returns>The service collection, to enable chaining.</returns>
+        public static IServiceCollection AddTenantedContentManagementApi(
+            this IServiceCollection services,
+            IConfiguration rootTenantDefaultConfiguration,
+            Action<IOpenApiHostConfiguration> configureHost = null)
+        {
+            Type contentServiceType = typeof(ContentService);
+
+            // If any of the OpenApi services are already installed, assume we've already completed installation and return.
+            if (services.Any(services => contentServiceType.IsAssignableFrom(services.ImplementationType)))
+            {
+                return services;
+            }
+
+            services.AddLogging();
+
+            services.AddContentManagementContent();
+
+            services.AddTenantCloudBlobContainerFactory(rootTenantDefaultConfiguration);
+            services.AddTenantProviderBlobStore();
+            services.AddTenantedAzureCosmosContentStore(rootTenantDefaultConfiguration);
+
+            services.AddContentManagementOpenApiServices();
+            services.AddContentManagementHalDocumentMappers();
+
+            services.AddOpenApiHttpRequestHosting<SimpleOpenApiContext>(config =>
+            {
+                config.AddContentManagementServiceDefinition();
+
+                configureHost?.Invoke(config);
+
+                config.MapTenancyExceptions();
+                config.MapContentManagementExceptions();
+            });
+
+            return services;
+        }
+
+        private static IServiceCollection AddContentManagementOpenApiServices(this IServiceCollection services)
+        {
+            services.AddSingleton<ContentService>();
+            services.AddSingleton<IOpenApiService, ContentService>(s => s.GetRequiredService<ContentService>());
+            return services;
+        }
+
+        private static IServiceCollection AddContentManagementHalDocumentMappers(this IServiceCollection services)
+        {
+            services.AddHalDocumentMapper<Content, IOpenApiContext, ContentMapper>();
+            return services;
+        }
+
+        private static void AddContentManagementServiceDefinition(this IOpenApiHostConfiguration config)
+        {
+            Type contentServiceType = typeof(ContentService);
```

Comment:
It's because this API has a large enough surface area that I want to split the implementation across multiple services.

Yes, I could attach the YAML to one of those via the attribute, but it just seems a bit... manky.

Reply:
Fair enough, in which case is there a good reason not to use the `OpenApiServiceDefinitions` class's `GetOpenApiServiceFromEmbeddedDefinition` or `ReadOpenApiServiceFromEmbeddedDefinitionWithDiagnostics` methods which were designed specifically to avoid having everyone copy and paste in the same boilerplate code for this scenario?

Reply:
And... you could always create partials for the different parts of the service and add the attribute to another partial.

Reply:
Ha, didn't know about `GetOpenApiServiceFromEmbeddedDefinition`.